src/wireviz/wv_bom.py

```diff
@@ -78,101 +101,91 @@ def generate_bom(harness):
-                else:
-                    designators.append(group_entry['designators'])
-        designators = list(dict.fromkeys(designators))  # remove duplicates
-        designators.sort()
-        total_qty = sum(entry['qty'] for entry in group_entries)
-        bom.append({**group_entries[0], 'qty': round(total_qty, 3), 'designators': designators})
-
-    bom = sorted(bom, key=lambda k: k['item'])  # sort list of dicts by their values (https://stackoverflow.com/a/73050)
-
-    # add an incrementing id to each bom item
-    bom = [{**entry, 'id': index} for index, entry in enumerate(bom, 1)]
-    return bom
-
-def get_bom_index(harness, item, unit, manufacturer, mpn, pn):
-    # Remove linebreaks and clean whitespace of values in search
-    target = tuple(clean_whitespace(v) for v in (item, unit, manufacturer, mpn, pn))
-    for entry in harness.bom():
-        if (entry['item'], entry['unit'], entry['manufacturer'], entry['mpn'], entry['pn']) == target:
+    for _, group in groupby(sorted(bom_entries, key=bom_entry_key), key=bom_entry_key):
+        group_entries = list(group)
+        designators = sum((make_list(entry.get('designators')) for entry in group_entries), [])
+        total_qty = sum(entry.get('qty', 1) for entry in group_entries)
+        bom.append({**group_entries[0], 'qty': round(total_qty, 3), 'designators': sorted(set(designators))})
+
+    # add an incrementing id to each bom entry
+    return [{**entry, 'id': index} for index, entry in enumerate(bom, 1)]
+
+def get_bom_index(bom: List[BOMEntry], target: BOMKey) -> int:
+    """Return id of BOM entry or raise exception if not found."""
+    for entry in bom:
+        if bom_entry_key(entry) == target:
             return entry['id']
-    return None
+    raise Exception('Internal error: No BOM entry found matching: ' + '|'.join(target))
 
-def bom_list(bom):
-    keys = ['id', 'item', 'qty', 'unit', 'designators'] # these BOM columns will always be included
-    for fieldname in ['pn', 'manufacturer', 'mpn']: # these optional BOM columns will only be included if at least one BOM item actually uses them
+def bom_list(bom: List[BOMEntry]) -> List[List[str]]:
+    """Return list of BOM rows as lists of column strings with headings in top row."""
+    keys = list(BOM_COLUMNS_ALWAYS)  # Always include this fixed set of BOM columns.
+    for fieldname in BOM_COLUMNS_OPTIONAL:  # Include only those optional BOM columns that are in use.
         if any(entry.get(fieldname) for entry in bom):
             keys.append(fieldname)
-    bom_list = []
-    # list of staic bom header names,  headers not specified here are generated by capitilising the internal name
+    # Custom mapping from internal name to BOM column headers.
+    # Headers not specified here are generated by capitilising the internal name.
     bom_headings = {
+        "description": "Item",  # TODO: Remove this line to use 'Description' in BOM heading.
```

Comment:
Consider removing this line and use `Description` instead of `Item` in BOM heading:
```suggestion
```
Originally, the term `item` was used both as one of the BOM column headers, and as any row in the BOM. Now, BOM `entry` is used as any row in the BOM, and `description` is used as internal name of the BOM column to be consistent with `additional_bom_items`.

If this suggestion is accepted, then this PR changes the output slightly, but only the BOM column header.

Reply:
However, if this TODO is considered controversial or otherwise might delay merging the PR, then please ignore it for now.